Add option to use prebuilt Electron from local

<!--
Thanks for opening an issue! A few things to keep in mind:

- The issue tracker is only for bugs and feature requests.
- Before reporting a bug, please try reproducing your issue against
  the latest version of Electron.
- If you need general advice, join our Slack: http://atom-slack.herokuapp.com
-->

* Electron version: Any
* Operating system: FreeBSD 11.1+ x64

### Expected behavior

<!-- What do you think should happen? -->

As part of the work on #3797, Electron dist/ can now be built and installed on FreeBSD. Since it is unlikely to get official support for FreeBSD from Electron team soon, it would be great if Electron can detect and use local dist when installed as a node_module of any Electron apps.

```
~ cd /usr/local/share/electron && ls -l
total 162125
-rw-r--r--  1 root  wheel       1060 Feb  6 20:29 LICENSE
-rw-r--r--  1 root  wheel    1788525 Mar 10 09:47 LICENSES.chromium.html
-rw-r--r--  1 root  wheel      25856 Mar 10 01:39 blink_image_resources_200_percent.pak
-rw-r--r--  1 root  wheel    7305848 Mar 22 15:31 chromedriver
-rw-r--r--  1 root  wheel    4033070 Mar 22 15:31 chromedriver-v1.8.2-freebsd-x64.zip
-rw-r--r--  1 root  wheel         15 Mar 10 01:39 content_resources_200_percent.pak
-rw-r--r--  1 root  wheel   10282427 Mar 10 01:50 content_shell.pak
-rwxr-xr-x  1 root  wheel  104013096 Mar 22 15:31 electron
-rw-r--r--  1 root  wheel     659579 Mar 22 15:31 electron-api.json
-rw-r--r--  1 root  wheel   57732748 Mar 22 15:31 electron-v1.8.2-freebsd-x64.zip
-rw-r--r--  1 root  wheel     355202 Mar 22 15:31 electron.d.ts
-rw-r--r--  1 root  wheel    1523849 Mar 22 15:31 ffmpeg-v1.8.2-freebsd-x64.zip
-rw-r--r--  1 root  wheel   10130560 Jun 26  2017 icudtl.dat
-rw-r--r--  1 root  wheel    2939664 Mar 22 15:31 libffmpeg.so
-rw-r--r--  1 root  wheel   24187344 Mar 22 15:31 libnode.so
drwxr-xr-x  2 root  wheel         56 Mar 22 15:32 locales
-rw-r--r--  1 root  wheel   11520672 Mar 22 15:31 mksnapshot
-rw-r--r--  1 root  wheel    5083066 Mar 22 15:31 mksnapshot-v1.8.2-freebsd-x64.zip
-rw-r--r--  1 root  wheel     239010 Mar 10 03:32 natives_blob.bin
-rw-r--r--  1 root  wheel     140979 Mar 22 15:21 pdf_viewer_resources.pak
drwxr-xr-x  2 root  wheel          4 Mar 22 15:32 resources
-rw-r--r--  1 root  wheel    1559008 Mar 10 04:01 snapshot_blob.bin
-rw-r--r--  1 root  wheel     151829 Mar 10 01:50 ui_resources_200_percent.pak
-rw-r--r--  1 root  wheel          6 Mar 22 15:31 version
-rw-r--r--  1 root  wheel      57761 Mar 10 01:50 views_resources_200_percent.pak
```

The installed content should be the same to downloaded archives (from `electron-vx.x.x-platform-x64.zip `).

### Actual behavior

<!-- What actually happens? -->

It has the following error since the prebuilt for FreeBSD is not exist:

```
> electron@1.8.4 postinstall /usr/home/guangyuan/cryptocat/node_modules/electron
> node install.js

Downloading electron-v1.8.4-freebsd-x64.zip
Error: GET https://github.com/electron/electron/releases/download/v1.8.4/electron-v1.8.4-freebsd-x64.zip returned 404
/usr/home/guangyuan/cryptocat/node_modules/electron/install.js:47
  throw err
  ^

Error: Failed to find Electron v1.8.4 for freebsd-x64 at https://github.com/electron/electron/releases/download/v1.8.4/electron-v1.8.4-freebsd-x64.zip
    at Request.<anonymous> (/usr/home/guangyuan/cryptocat/node_modules/nugget/index.js:169:61)
    at Request.emit (events.js:159:13)
    at Request.onRequestResponse (/usr/home/guangyuan/cryptocat/node_modules/request/request.js:1068:10)
    at ClientRequest.emit (events.js:159:13)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:550:21)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:115:23)
    at TLSSocket.socketOnData (_http_client.js:439:20)
    at TLSSocket.emit (events.js:159:13)
    at addChunk (_stream_readable.js:265:12)
    at readableAddChunk (_stream_readable.js:252:11)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! electron@1.8.4 postinstall: `node install.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the electron@1.8.4 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/guangyuan/.npm/_logs/2018-03-22T19_35_18_972Z-debug.log
```

<!-- 

Your best chance of getting this bug looked at quickly is to provide a REPOSITORY that can be cloned and run.

You can fork https://github.com/electron/electron-quick-start and include a link to the branch with your changes.

If you provide a URL, please list the commands required to clone/setup/run your repo e.g.

  $ git clone $YOUR_URL -b $BRANCH
  $ npm install
  $ npm start || electron .

-->

--- a/.gitignore
+++ b/.gitignore
@@ -43,3 +43,8 @@
 node_modules/
 SHASUMS256.txt
 **/package-lock.json
+**/yarn.lock
+
+# npm package
+/npm/dist
+/npm/path.txt

--- a/docs/api/environment-variables.md
+++ b/docs/api/environment-variables.md
@@ -88,10 +88,10 @@ This environment variable will not work if the `crashReporter` is started.
 
 ### `ELECTRON_OVERRIDE_DIST_PATH`
 
-When running from the `electron` package using this variable will point
-the `electron` command at a locally built version of Electron instead of
-the one that was downloading during `npm install`.  Usage is like:
+When running from the `electron` package, this variable tells
+the `electron` command to use the specified build of Electron instead of
+the one downloaded by `npm install`. Usage:
 
-```bash
+```sh
 export ELECTRON_OVERRIDE_DIST_PATH=/Users/username/projects/electron/out/D
 ```

--- a/docs/api/environment-variables.md
+++ b/docs/api/environment-variables.md
@@ -85,3 +85,13 @@ This environment variable will not work if the `crashReporter` is started.
 Shows the Windows's crash dialog when Electron crashes.
 
 This environment variable will not work if the `crashReporter` is started.
+
+### `ELECTRON_OVERRIDE_DIST_PATH`
+
+When running from the `electron` package using this variable will point
+the `electron` command at a locally built version of Electron instead of
+the one that was downloading during `npm install`.  Usage is like:
+
+```bash
+export ELECTRON_OVERRIDE_DIST_PATH=/Users/username/projects/electron/out/D
+```

--- a/npm/index.js
+++ b/npm/index.js
@@ -3,7 +3,7 @@ var path = require('path')
 
 var pathFile = path.join(__dirname, 'path.txt')
 
-function getElectronPath() {
+function getElectronPath () {
   if (fs.existsSync(pathFile)) {
     var executablePath = fs.readFileSync(pathFile, 'utf-8')
     if (process.env.ELECTRON_OVERRIDE_DIST_PATH) {

--- a/npm/index.js
+++ b/npm/index.js
@@ -3,8 +3,16 @@ var path = require('path')
 
 var pathFile = path.join(__dirname, 'path.txt')
 
-if (fs.existsSync(pathFile)) {
-  module.exports = path.join(__dirname, fs.readFileSync(pathFile, 'utf-8'))
-} else {
-  throw new Error('Electron failed to install correctly, please delete node_modules/electron and try installing again')
+function getElectronPath() {
+  if (fs.existsSync(pathFile)) {
+    var executablePath = fs.readFileSync(pathFile, 'utf-8')
+    if (process.env.ELECTRON_OVERRIDE_DIST_PATH) {
+      return path.join(process.env.ELECTRON_OVERRIDE_DIST_PATH, executablePath)
+    }
+    return path.join(__dirname, 'dist', executablePath)
+  } else {
+    throw new Error('Electron failed to install correctly, please delete node_modules/electron and try installing again')
+  }
 }
+
+module.exports = getElectronPath()

--- a/npm/install.js
+++ b/npm/install.js
@@ -52,12 +52,12 @@ function getPlatformPath () {
 
   switch (platform) {
     case 'darwin':
-      return 'dist/Electron.app/Contents/MacOS/Electron'
+      return 'Electron.app/Contents/MacOS/Electron'
     case 'freebsd':
     case 'linux':
-      return 'dist/electron'
+      return 'electron'
     case 'win32':
-      return 'dist/electron.exe'
+      return 'electron.exe'
     default:
       throw new Error('Electron builds are not available on platform: ' + platform)
   }

